Mathlib/Data/Ordmap/Ordnode.lean

```diff
@@ -0,0 +1,1365 @@
+    if sl > sr then
+      let (l', m) := splitMax' ll lx lr
+      balanceR l' m r
+    else
+      let (m, r') := splitMin' rl rx rr
+      balanceL l m r'
+#align ordnode.glue Ordnode.glue
+
+/-- O(log(m + n)). Concatenate two trees that are ordered with respect to each other.
+
+     merge {1, 2} {3, 4} = {1, 2, 3, 4}
+     merge {3, 4} {1, 2} = precondition violation -/
+def merge (l : Ordnode α) : Ordnode α → Ordnode α :=
+  -- Porting note: Previous code was:
+  -- (Ordnode.recOn l fun r => r) fun ls ll lx lr IHll IHlr r =>
+  --   (Ordnode.recOn r (node ls ll lx lr)) fun rs rl rx rr IHrl IHrr =>
+  --     if delta * ls < rs then balanceL IHrl rx rr
+  --     else
+  --       if delta * rs < ls then balanceR ll lx (IHlr <| node rs rl rx rr)
+  --       else glue (node ls ll lx lr) (node rs rl rx rr)
+  --
+  -- failed to elaborate eliminator, expected type is not available
+  match l with
+  | nil => fun x ↦ x
+  | node ls ll lx lr => fun r ↦
+    match r with
+    | nil => node ls ll lx lr
+    | node rs rl rx rr =>
+      if delta * ls < rs then balanceL (merge ll r) rx rr
+      else if delta * rs < ls then balanceR ll lx (merge lr <| node rs rl rx rr)
+      else glue (node ls ll lx lr) (node rs rl rx rr)
+#align ordnode.merge Ordnode.merge
+
+/-- O(log n). Insert an element above all the others, without any comparisons.
+(Assumes that the element is in fact above all the others).
+
+    insertMax {1, 2} 4 = {1, 2, 4}
+    insertMax {1, 2} 0 = precondition violation -/
+def insertMax : Ordnode α → α → Ordnode α
+  | nil, x => ι x
+  | node _ l y r, x => balanceR l y (insertMax r x)
+#align ordnode.insert_max Ordnode.insertMax
+
+/-- O(log n). Insert an element below all the others, without any comparisons.
+(Assumes that the element is in fact below all the others).
+
+    insertMin {1, 2} 0 = {0, 1, 2}
+    insertMin {1, 2} 4 = precondition violation -/
+def insertMin (x : α) : Ordnode α → Ordnode α
+  | nil => ι x
+  | node _ l y r => balanceR (insertMin x l) y r
+#align ordnode.insert_min Ordnode.insertMin
+
+/-- O(log(m+n)). Build a tree from an element between two trees, without any
+assumption on the relative sizes.
+
+    link {1, 2} 4 {5, 6} = {1, 2, 4, 5, 6}
+    link {1, 3} 2 {5} = precondition violation -/
+def link (l : Ordnode α) (x : α) : Ordnode α → Ordnode α :=
+  -- Porting note: Previous code was:
+  -- (Ordnode.recOn l (insertMin x)) fun ls ll lx lr IHll IHlr r =>
+  --   (Ordnode.recOn r (insertMax l x)) fun rs rl rx rr IHrl IHrr =>
+  --     if delta * ls < rs then balanceL IHrl rx rr
+  --     else if delta * rs < ls then balanceR ll lx (IHlr r) else node' l x r
+  --
+  -- failed to elaborate eliminator, expected type is not available.
+  match l with
+  | nil => insertMin x
+  | node ls ll lx lr => fun r ↦
+    match r with
+    | nil => insertMax l x
+    | node rs rl rx rr =>
+      if delta * ls < rs then balanceL (link ll x rl) rx rr
+      else if delta * rs < ls then balanceR ll lx (link lr x rr)
```

Comment:
This now builds, after https://github.com/leanprover-community/mathlib4/pull/1455/commits/66117186a33c250b427f3b854db0d99547df1d9e, but I don't know if this is really what we want. 
Is there a way to know what exactly `IHrl` and `IHlr` were in mathlib?
Any ideas, @qawbecrdtey?

Reply:
The code is about balancing a tree which looks like
```text
  root
   /  \
  x  rr
 / \
l   rl
```
or its symmetric version.
So I believe it's correct.